### PR TITLE
feat(OpenStreetMap): new presence

### PR DESCRIPTION
**Pull request by Lulucmy**
## Description 
Adding the presence of OpenStreetMap, a collaborative mapping project.
Huge thanks to @DarkVillager and @kyrie25 for their help!

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

</details>

### CODE REVIEW

File: websites/O/OpenStreetMap/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+	"$schema": "https://schemas.premid.app/metadata/1.7",
+	"author": {
+		"name": "frnk",
+		"id": "232904284004548608"
+	},
+	"service": "OpenStreetMap",
+	"description": {
+		"en": "OpenStreetMap is a collaborative online mapping project where every user can contribute to the map.",
+		"fr": "OpenStreetMap est un projet de cartographie collaboratif en ligne où chaque utilisateur peut contribuer à la carte."
```
Comment (EncryptedDev) [inline websites/O/OpenStreetMap/dist/metadata.json:10]: @Kozou4ever

File: websites/O/OpenStreetMap/presence.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	if (x && y) {
+		fetch(
+			`https://nominatim.openstreetmap.org/reverse?lat=${x}&lon=${y}&format=json`
+		)
+			.then(res => res.json())
+			.then(res => {
+				locationcity = res.address.city;
+				if (!locationcity) locationcity = res.address.hamlet;
+				if (!locationcity) locationcity = res.address.village;
+				if (!locationcity) locationcity = res.address.town;
```
Comment (EncryptedDev) [inline websites/O/OpenStreetMap/presence.ts:23]: Use [nullish coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)

File: websites/O/OpenStreetMap/presence.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+		presenceData.state = "GeoDNS Zones";
+	else if (document.location.hostname.startsWith("git"))
+		presenceData.state = "Project Git";
+	else if (document.location.hostname.startsWith("lists"))
+		presenceData.state = "Project Mailing Lists";
+	else if (document.location.hostname.startsWith("blogs"))
+		presenceData.state = "Community blogs";
+	else if (document.location.hostname.startsWith("help")) {
+		presenceData.details = "Reading help:";
+		presenceData.state = document.title.split(" -")[0];
```
Comment (EncryptedDev) [inline websites/O/OpenStreetMap/presence.ts:83]: Use the Record method here as shown below instead
https://github.com/PreMiD/Presences/blob/113adabd2296ec1be22122ccc3ec2deff1c399ed/websites/T/TruckersMP/presence.ts#L15-L41

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
This presence is quite privacy invasive. I would add a privacy setting and implement this through the code too. Maybe display country only and buttons with the privacy setting true

**Review by itsmelouis — APPROVED** [review]

File: websites/O/OpenStreetMap/dist/metadata.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	},
+	"url": "openstreetmap.org",
+	"regExp": "([a-z0-9]+)[.]openstreetmap[.]org",
+	"version": "1.0.0",
+	"logo": "https://i.imgur.com/mlLReW9.png",
+	"thumbnail": "https://i.imgur.com/uXANhtt.png",
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/dist/metadata.json:16]: why paris? Couldn't you put a picture of the entire world

File: websites/O/OpenStreetMap/presence.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		)
+			.then(res => res.json())
+			.then(res => {
+				locationcity = res.address.city;
+				if (!locationcity) locationcity = res.address.hamlet;
+				if (!locationcity) locationcity = res.address.village;
+				if (!locationcity) locationcity = res.address.town;
+
+				locationstate = res.address.state;
+				locationcountry = res.address.country;
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:26]: use object destructuring

File: websites/O/OpenStreetMap/presence.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	y: string,
+	editedObject: string,
+	selectedObject: boolean;
+setInterval(() => {
+	if (x && y) {
+		fetch(
+			`https://nominatim.openstreetmap.org/reverse?lat=${x}&lon=${y}&format=json`
+		)
+			.then(res => res.json())
+			.then(res => {
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:19]: you can keep the .then above this line but for this callback please use async/await instead

File: websites/O/OpenStreetMap/presence.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+			})
+			.catch();
+	}
+}, 5000);
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+		largeImageKey: "logo",
+		startTimestamp: browsingTimestamp,
+		details: "Current page:"
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:36]: you shouldn't do this. Just display each page's data in the details, maybe change the strings a little to say "Viewing" or whatever you'd like but don't do this

File: websites/O/OpenStreetMap/presence.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	};
+
+	if (document.location.pathname.includes("/traces/new"))
+		presenceData.state = "GPS tracks upload";
+	else if (document.location.pathname.includes("/traces"))
+		presenceData.state = "Public GPS tracks";
+	else if (document.location.pathname.includes("/diary/new"))
+		presenceData.state = "Write new diary entry";
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:44]:
```suggestion
		presenceData.state = "Writing a new diary entry";
```

File: websites/O/OpenStreetMap/presence.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	const presenceData: PresenceData = {
+		largeImageKey: "logo",
+		startTimestamp: browsingTimestamp,
+		details: "Current page:"
+	};
+
+	if (document.location.pathname.includes("/traces/new"))
+		presenceData.state = "GPS tracks upload";
+	else if (document.location.pathname.includes("/traces"))
+		presenceData.state = "Public GPS tracks";
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:42]: what's a GPS track?

File: websites/O/OpenStreetMap/presence.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	else if (document.location.hostname.startsWith("lists"))
+		presenceData.state = "Project Mailing Lists";
+	else if (document.location.hostname.startsWith("blogs"))
+		presenceData.state = "Community blogs";
+	else if (document.location.hostname.startsWith("help")) {
+		presenceData.details = "Reading help:";
+		presenceData.state = document.title.split(" -")[0];
+	} else if (document.location.hostname.startsWith("blog")) {
+		presenceData.details = "Reading a blog post:";
+		presenceData.state = document.title.split(" |")[0];
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:86]: I know it works the same, but why the space
```suggestion
		presenceData.state = document.title.split("|")[0];
```

File: websites/O/OpenStreetMap/presence.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+		presenceData.buttons = [
+			{ label: "Read this post", url: document.location.href }
+		];
+	} else if (document.location.hostname.startsWith("forum")) {
+		presenceData.details = "Reading a forum post:";
+		presenceData.state = document.title.split(" /")[0];
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:92]:
```suggestion
		presenceData.state = document.title.split("/")[0];
```

File: websites/O/OpenStreetMap/presence.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+		presenceData.buttons = [
+			{ label: "Read this post", url: document.location.href }
+		];
+	} else if (document.location.hostname.startsWith("community")) {
+		presenceData.details = "Reading a community post:";
+		presenceData.state = document.title.split(" -")[0];
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:98]:
```suggestion
		presenceData.state = document.title.split("-")[0];
```

File: websites/O/OpenStreetMap/presence.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+		presenceData.buttons = [
+			{ label: "Read this post", url: document.location.href }
+		];
+	} else if (document.location.pathname.includes("/search")) {
+		presenceData.details = "Search results for:";
+		presenceData.state = document.title.split("|")[0];
+	} else if (document.location.pathname.includes("/user/")) {
+		presenceData.details = `On ${document.title.split("|")[0]} profile :`;
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:106]:
```suggestion
		presenceData.details = `On ${document.title.split("|")[0]}'s profile:`;
```

File: websites/O/OpenStreetMap/presence.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+		if (typeof locationcountry == "undefined")
+			presenceData.state = "Somewhere on planet earth";
+		else
+			presenceData.state = `Near ${locationcity}, ${locationstate}, ${locationcountry}`;
+		presenceData.buttons = [
+			{ label: "Open the map", url: document.location.href }
+		];
+	} else if (document.location.pathname.startsWith("/wiki")) {
+		presenceData.details = "Reading the wiki";
+		presenceData.state = `"${document.title.split(" —")[0]}"`;
```
Comment (ImRodry) [inline websites/O/OpenStreetMap/presence.ts:194]:
```suggestion
		presenceData.state = `"${document.title.split("—")[0]}"`;
```

**Review by ImRodry — CHANGES_REQUESTED** [review]

**Comment by ImRodry** [comment]
Rebased to apply the latest changes in the prettier configuration. If you need help with your local branch, please ping me on the #presence-dev channel on Discord.

**Review by Lulucmy — COMMENTED** [review]
Sorry for the delay, I'm busy IRL but I'll keep working on it

**Comment by ImRodry** [comment]
Please address the unresolved conversations in a timely manner, otherwise this PR will be closed

**Comment by EncryptedDev** [comment]
Inactive

**Comment by EncryptedDev** [comment]
Resolve CI

**Comment by Lulucmy** [comment]
Should be good now 👍

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
This is still unresolved https://github.com/PreMiD/Presences/pull/6305/files#r895082509

**Review by EncryptedDev — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**